### PR TITLE
Coalesce haproxy reloads with a configurable interval (default 15 secs).

### PR DESCRIPTION
**Pull request by ramr**
@rajatchopra   PTAL.  Thx

### CODE REVIEW

**Comment by ramr** [comment]
@eparis  since Rajat is out - can you PTAL. Thx

File: images/router/haproxy/reload-haproxy [hunk 1]
```diff
@@ -1,21 +1,64 @@
 #!/bin/bash -xu
 
+script=$(basename "${BASH_SOURCE[0]}")
+DEFAULT_COALESCE_INTERVAL=15   #  default 15 secs.
+COALESCE_MAX_WAIT=$((5 * 60))  #  5 minutes
```
Comment (eparis) [inline images/router/haproxy/reload-haproxy]: shouldn't these be lower case instead of upper case?

Comment (ramr) [inline images/router/haproxy/reload-haproxy]: Could use either  - just added it to indicate a constant (though that said all the other older stuff below is in lower case). So if you want me to, I can change it to lower case. 
`BASH_SOURCE` however needs to be uppercase - its the script/source name (from bash).

Comment (eparis) [inline images/router/haproxy/reload-haproxy]: Its up to you, I think the convention is lowercase for variable which are not from the outside or exposed to the outside (although in this case they are global, so you can't use local)   it's totally style....

Comment (ramr) [inline images/router/haproxy/reload-haproxy]: I'll update it.

**Comment by ramr** [comment]
updated - as mentioned on irc - made  those constants an int and readonly.

**Comment by eparis** [comment]
Lets say reload-haproxy is called at time 0.
At time 15.1 lets say the haproxy reload starts.
At time 15.2 another call to reload-haproxy is made. It will find the flock held and will exit.
At time 15.3 the original haproxy reload may return and the process exit

There is now a change in the config and no timer to update...

**Comment by eparis** [comment]
is any of the flock -u stuff required? they'll get cleaned up when you exit....

**Comment by ramr** [comment]
Yeah, that's true.  The lock is held for the time it takes to launch the haproxy process (not the time it takes for process the config - should be a couple of ms odd). On a loaded environment, ala online since there would be routes being constantly added, am assuming this won't be hit. I was exiting because I wanted to exit rather than fire off a background reload (as that does mean we do more reloads than are required). 

But you are correct in that there is a possibility of an edge case where on a loaded cpu, the config might just be updated past the haproxy reload (though that said haproxy would need to hog the cpu in order for that to happen).

Let me try and fix that differently - maybe match config timestamps to a file touched just before the reload and do the reload if needed in the background nohupped. That way, we end up doing not as many reloads for the hosted/online case.  Let me mull on this a bit more. 

The `flock -u` stuff I had a few issues without having that in place so added it in for safety's sake.

**Comment by ramr** [comment]
@eparis  I reworked this to address the review comments + possible error cases  w/ a separate lock for coalescing the reloads and a reload lock to ensure only 1 haproxy reload runs at a time (fyi, the current code had this issue too so its good to fix that as well).  Tested this running a series of reloads 
```#!/bin/bash  
  
while true; do ./reload-haproxy ; sleep $((RANDOM % 5)); done  
```
and 10 of these scripts nohupped (`nohup ./t1.sh &> /dev/null &`) running in parallel and it works well. 

PTAL  Thx

File: images/router/haproxy/reload-haproxy [hunk 2]
```diff
@@ -1,21 +1,98 @@
 #!/bin/bash -xu
```
Comment (eparis) [inline images/router/haproxy/reload-haproxy]: Do we need -x ?

**Comment by eparis** [comment]
Can we write a real daemon which we signal to do this? Because even this isn't 'right'.

Between the time you `rmdir "$coalesce_lock"` and start the subshell inside `lock_and_reload_haproxy` which touches `$reload_lock` any number of other calls to `reload-haproxy` can race past `coalesce_reloads` and cause multiple reloads in very fast succession.

But I'd much rather have it like this than the race lose a reload.

I'm also concerned about those sorts being safe against whatever fills in those files. I know that has nothing to do with this PR just my little brain being too small.

So LGTM (aside from the -x) but if I can write a real daemon I'd like it even better, don't know how reasonable that is though since I don't know the rest of the pod...

File: images/router/haproxy/reload-haproxy (hunk 2 above)
Comment (ramr) [inline images/router/haproxy/reload-haproxy]: It was there from before and I sort like it (it helps with debugging - tells you what the script did/is doing).

**Comment by ramr** [comment]
@eparis  will see what I can do vis-a-vis a daemon over my break. For now, this provides that support. Yeah, so the  earlier version was aggressively reducing the number of reloads, this one is gating it and so a flurry of reloads may well cause more reloads than we want but in a stress test running 11 scripts (10 backgrounded and 1 in the foreground) I actually saw it behaving decently, so ...

**Comment by eparis** [comment]
I'll write the daemon, something to do in my free moments. I just don't know how to signal it as needed or if that will be a huge PITA. You tell me if I should bother, this PR is LGTM as it stands.

**Comment by ramr** [comment]
don't worry about it - I'll do it - think there could be a simpler `gating` and locking mechanism we can use.

**Comment by spinolacastro** [comment]
I've seen a lot of haproxy restarts on my cluster and my mainly concern is about connections being refused between restarts.

What do you think of Queuing Disciplines or Iptables to avoid this behavior ?

<pre>

iptables -A INPUT -p tcp --dports 80,443 -j DROP

haproxy_restart()

iptables -D INPUT -p tcp --dports 80,443 -j DROP

</pre>

As far as i could dig, we have to run the router pod with special caps.

**Comment by eparis** [comment]
@knobunc has looked at doing something like that. I'll let him link to details.

**Comment by spinolacastro** [comment]
@eparis sorry, i just scrolled the page and found the PR (https://github.com/openshift/origin/pull/6472) :)

**Comment by smarterclayton** [comment]
Why isn't this in go code?  Go code is a lot better at managing this, and it benefits all possible router types.

**Comment by smarterclayton** [comment]
Specifically, openshift-router is already a daemon hat manages this process

**Comment by ramr** [comment]
@smarterclayton  was going to move it into the infra router code since it was getting more complicated in here. Will do that when am i back tomorrow. Though we might still an additional check in the shell script to reload - will look into it tmrw.

**Pull request closed without merging**